File: nearest-pairs/pairs.erl
%% Return the two closest pairs of a list of coordinates

-module(pairs).
-export([squaredist/1, dist/1, pairs/1, closest/1, pairdist/1]).

%% Return the square of the distance between two points
squaredist([{A1, A2}, {B1, B2}]) ->
    math:pow((B1-A1), 2)+ math:pow((B2-A2), 2).

dist([{A1, A2}, {B1, B2}]) ->
    math:sqrt(squaredist([{A1, A2}, {B1, B2}])).

pairs([_]) ->
    [];
pairs([A, B]) ->
    [A | B];
pairs([H | T]) ->
    [[H | El] || El <- T] ++ pairs(T).

pairdist([_]) ->
    [];
pairdist([A, B]) ->
    [{pair, A, B, squaredist([A, B])}];
pairdist([H | T]) ->
    [{pair, H, El, squaredist([H, El])} || El <- T] ++ pairdist(T).

File: nearest-pairs/pairs.py
"""
Find closest pairs of a list of coordinates
"""

from itertools import combinations
from math import sqrt

def squaredist(P1, P2):
    return pow(P1[0]-P2[0], 2) + pow(P1[1]-P2[1], 2)

def dist(P1, P2):
    return sqrt(squaredist(P1, P2))

def closest(coordinates):
    (closest, shortest_dist) = (None, 0)
    for (P1, P2) in combinations(coordinates, 2):
        sqdist = squaredist(P1, P2)
        if closest is None or sqdist < shortest_dist:
            shortest_dist = sqdist
            closest = (P1, P2)
    return closest

def main():
    tests = (
        [(0, 0), (1, 20), (5, 2)],
        [(-10, 10), (1, 5), (4, 3)]
    )
    for testset in tests:
        print "The closest points of ", testset, " is ", closest(testset)
    
if __name__ == "__main__":
    main()

